cpp/include/partition_manager.hpp

```diff
@@ -0,0 +1,171 @@
+// responsible with finding appropriate P_ROW x P_COL
+// 2D partition and initializing the raft::handle_t communicator
+//
+// (this might be removed; or, it might exist already)
+//
+template <typename size_type = int>
+class partition_manager_t {
+ public:
+  partition_manager_t(raft::handle_t& handle, size_type p_row_size, size_type p_col_size)
+    : handle_(handle), p_row_size_(p_row_size), p_col_size_(p_col_size)
+  {
+    init_communicator();
+  }
+
+  partition_manager_t(raft::handle_t const& handle, size_type p_size) : handle_(handle)
+  {
+    partition2d(p_size);
+    init_communicator();
+  }
+
+  virtual ~partition_manager_t(void) {}
+
+ protected:
+  virtual void partition2d(size_type p_size)
+  {
+    auto sqr = static_cast<size_type>(std::sqrt(p_size));
+
+    // find divisor of p_size
+    // nearest to sqr;
+    //
+    p_row_size_ = nearest_divisor(sqr, p_size);
+    p_col_size_ = p_size / p_row_size_;
+
+    assert(p_row_size_ > 1 && p_col_size_ > 1);
+  }
+
+  virtual void init_communicator(void)
+  {
+    // TODO: init's handle's communicator (singleton?)
+  }
+
+ private:
+  raft::handle_t& handle_;
+  size_type p_row_size_;
+  size_type p_col_size_;
+
+  static decltype(auto) nearest_divisor(size_type sqr, size_type p_size)
+  {
+    assert(sqr > 0);
+
+    for (size_type div = sqr; div > 0; --div) {
+      auto p_div = p_size % div;
+      if (p_div == 0) return div;
+    }
+  }
+};
+
+// default key-naming mechanism:
+//
+struct key_naming_t {
+  key_naming_t(int row_indx,
+               int col_indx,
+               std::string const& col_suffix = std::string("_col"),
+               std::string const& row_suffix = std::string("_row"),
+               std::string const& prefix     = std::string("partition"))
+    : col_suffix_(col_suffix),
+      row_suffix_(row_suffix),
+      prefix_(prefix),
+      name_(prefix_ + "_" + to_string(row_indx) + "_" + to_string(col_indx))
+  {
+  }
+
+  std::string col_name(void) const { return name_ + col_suffix_; }
+
+  std::string row_name(void) const { return name_ + row_suffix_; }
+
+ private:
+  std::string const col_suffix_;
+  std::string const row_suffix_;
+  std::string const prefix_;
+  std::string name_;
+};
+
+using pair_comms_t =
+  std::pair<std::shared_ptr<raft::comms::comms_t>, std::shared_ptr<raft::comms::comms_t>>;
+
+enum class key_2d_t : int { ROW = 0, COL = 1 };
+
+// class responsible for creating 2D partition sub-comms:
+// this is instantiated by each worker (processing element, PE)
+// for the row/column it belongs to;
+//
+template <typename name_policy_t = key_naming_t, typename size_type = int>
+class subcomm_factory_t {
+ public:
+  subcomm_factory_t(raft::handle_t& handle, size_type p_row_index, size_type p_col_index)
+    : handle_(handle), row_index_(p_row_index), col_index_(p_col_index)
+  {
+    init_row_col_comms();
+  }
+  virtual ~subcomm_factory_t(void) {}
+
+ protected:
+  virtual void init_row_col_comms(void)
```

Comment:
I think what we need is just this function. Currently we're calling only set_comms() on handle. We need to call this function as well. Talk with @iroy30 about the best place to call this function.

Reply:
Yes, this is the central piece of functionality. I will discuss this further with @Iroy30. Thanks.

Reply:
@Iroy30, so where should this function called from? Thanks in advance.

Reply:
@seunghwak When does this subcomm creation need to happen? I was thinking the cython wrapper right where set_comms is called , since it needs to be called by each process and we would just pass the handle and row_size to the subcomm_factory_t call their before initializing the 2D graph

Reply:
@Iroy30 Yes, this needs to be called when we call set_comms. We set-up the global communicator first, then right after that, we set-up the sub-communicators. We put more control functions in the python layer, and if we call this function after setting up-the global communicator, I think most code outside this initi_row_col_comms function becomes irrelevant.

Reply:
So, what is the action item for me, here? Where do I initialize the `subcomm_factory_t ` object?

If this is not yet clear (or the ideal location is "under construction") perhaps defer this to a follow-up PR and merge this as is?

Reply:
Yes this just needs to be called in cython before calling the mnmg algorithms with 2D partitioning. The call should be included/updated in a separate mnmg pipeline PR. This PR can be merged as is.

Reply:
OK, resolving this. If this can be merged, can I get enough approvals to have it mergeable? (I'll handle conflicts shortly). Thanks in advance.